rfd/0123-tls-routing-behind-layer7-lb.md

```diff
@@ -228,6 +228,28 @@ See [RFD 0121 - Kubernetes MFA
 sessions](https://github.com/gravitational/teleport/blob/master/rfd/0121-kube-mfa-sessions.md)
 for more details.
 
+### Client source IPs
+
+Correct client source IPs must be reported for IP pinning and audit logs.
+
+Unlike layer 4 load balancers, layer 7 load balancers usually CANNOT transparently forward original clients' IPs or add
+standard [PROXY](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) protocol headers to layer 7 payload. The
+standard way for layer 7 load balancers to report correct client source IPs is through the
+["X-Forwarded-For"](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For) (XFF) header.
```

Comment:
`X-Forward-For` is a challenging header to use.  Proxy implementations commonly append to this rather than replace the value.  That can open up a frequently seen attack where a malicious client provides in the request a false IP, leaving their actually IP ignored within the chain provided.

In my experience `X-Real-IP` (which is only a single value) is simpler and less likely to be implemented incorrectly.  Did you consider using this header instead?

Reply:
> In my experience X-Real-IP (which is only a single value) is simpler and less likely to be implemented incorrectly. Did you consider using this header instead?

It depends if `X-Real-IP` is widely supported. AWS ALB for example, one of the primary use cases for this feature, only supports `X-Forwarded` headers. I think same applies to HAProxy.

> X-Forward-For is a challenging header to use. Proxy implementations commonly append to this rather than replace the value

Yes, hence this proposal to use last IP from the header value. What do you think?

Reply:
>  I think same applies to HAProxy

I believe `haproxy` supports proxy protocol: https://access.redhat.com/solutions/6888131

> use last IP from the header value, if there is a single XFF header value. Multiple values will be rejected.

For AWS ALB's this is a safe assumption since they will reliably put the client ip in the left position.

With `haproxy` this may also be safe, but it seems may result in being non-functional when multiple `haproxy` instances are used in series.  Reading the docs here: http://docs.haproxy.org/dev/configuration.html#option%20forwardfor

Relevant Quote:
```
Since this header is always appended at the end of the existing header list, the server must be configured to 
always use the last occurrence of this header only.
```
I understand this to mean that multiple haproxy instances in series will result in multiple instances of this header.  And that the order of the headers are how they ensure spoofing is protected against.

In general our users would just need to be certain that their proxies replace the header at the edge, and any internal proxies they may have append on the left.  The need for a different configuration across systems can be challenging for some clients.  But I agree that `X-Real-IP` is less widely supported, so this is probably best addressed in documentation.

Reply:
> I believe haproxy supports proxy protocol: https://access.redhat.com/solutions/6888131

Do you know if PROXY protocol can be used for HTTP mode (layer 7) for `haproxy`? Looks like there is no restriction on that. I think that's better than using this header. AWS NLB can do PROXY protocol but not ALB.

> In general our users would just need to be certain that their proxies replace the header at the edge, and any internal proxies they may have append on the left. 

Proposed in this rfd to use a config `proxy_service.use_xff_header` (naming might change) to toggle intake of XFF value. Only when the user is certain about their environment setup, set this config to true then Teleport will trust the last IP in the single XFF header. Any risk you see this way? or please advice you see a better approach.

thanks!